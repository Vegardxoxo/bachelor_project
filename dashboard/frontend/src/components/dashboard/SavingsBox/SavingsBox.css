#title {
    color: #fff;
    padding-top: 5px;
    padding-left: 5px;
    text-align: center;
    font-family: Source Sans Pro, sans-serif;
    font-size: 20pt;
    font-weight: 400;
}

#numbers {
    color: #fff;
    padding-left: 15px;
    padding-top: 10px;
    text-align: left;
    font-family: Source Sans Pro, sans-serif;
    font-size: 35px;
    font-weight: lighter;
}