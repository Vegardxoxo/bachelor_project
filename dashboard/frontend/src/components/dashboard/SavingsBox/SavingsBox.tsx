import './SavingsBox.css';
import React, {useEffect, useState} from 'react';
import {Card, CardContent, Stack, Tooltip, Typography} from '@mui/material';
import CardOverflow from '@mui/joy/CardOverflow';
import {orgAtom} from '../../../globalVariables/variables';
import {useRecoilValue} from 'recoil';
import {fetchPotentialSavings} from '../../../api/calls';

/**
 * Returns a box displaying the potential savings for the organization.
 */
export function SavingsBox() {

    // Declare and initiate state variables
    const [potentialSavings, setpotentialSavings] = useState<number | undefined>(undefined);
    const organization = useRecoilValue(orgAtom)

    // Fetch data from API when component mounts or access token/organization changes
    useEffect(() => {
        const fetchData = async () => {
            if (organization) {
                try {
                    const data: string[] | undefined = await fetchPotentialSavings(organization);
                    if (data !== undefined) {
                        const numberData: number = +data;
                        setpotentialSavings(numberData);
                    }
                } catch (error) {
                    console.error('Error fetching potential savings:', error);
                }
            }
        };

        fetchData();
    }, [organization]);

    /* Function returns a SavingsBox card with information on potential savings fetched from the API */
    return (
        <Tooltip title={<h2 style={{fontSize: 15, fontWeight: 'lighter'}}>
            Potensiell sparing viser hvor mye enheten kan spare ved Ã¥ frigi alle ledige og ubrukte lisenser.
        </h2>} followCursor arrow placement='top'>
            <Card sx={{width: 300, height: 140, borderRadius: 5, backgroundColor: '#002d53'}} data-testid='savingsBox'>
                <CardOverflow>
                    <CardContent>
                        <Stack direction={'row'}>
                            <Typography id="title">
                                Potensiell sparing
                            </Typography>
                        </Stack>
                        <Typography id="numbers">
                            {potentialSavings?.toLocaleString('nb-NO', {useGrouping: true})} kr
                        </Typography>
                    </CardContent>
                </CardOverflow>
            </Card>
        </Tooltip>
    )
}