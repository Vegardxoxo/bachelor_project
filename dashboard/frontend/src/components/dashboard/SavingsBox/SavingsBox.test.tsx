import {cleanup, fireEvent, render, screen} from '@testing-library/react';
import {SavingsBox} from './SavingsBox';
import '@testing-library/jest-dom/extend-expect';
import {MemoryRouter} from 'react-router-dom';
import {RecoilRoot} from 'recoil';
import renderer from "react-test-renderer";
import React from "react";


describe('Testing render, DonutChart', () => {

    beforeEach(() => {
        render(
            <MemoryRouter>
                <RecoilRoot>
                    <SavingsBox/>
                </RecoilRoot>
            </MemoryRouter>);
    });

    afterEach(() => {
        cleanup();
    });

    it('renders donutchart without crashing', () => {
        expect(screen.getByText(/Potensiell sparing/)).toBeInTheDocument();
        expect(screen.getByTestId('savingsBox')).toBeInTheDocument();
    });

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <MemoryRouter>
                <RecoilRoot>
                    <SavingsBox/>
                </RecoilRoot>
            </MemoryRouter>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })


    it('renders tooltip when hovering over SavingsBox', async () => {
        fireEvent.mouseEnter(screen.getByTestId('savingsBox'))
        await screen.findByRole(/tooltip/);
        expect(screen.getByRole(/tooltip/)).toBeInTheDocument();
    });

});