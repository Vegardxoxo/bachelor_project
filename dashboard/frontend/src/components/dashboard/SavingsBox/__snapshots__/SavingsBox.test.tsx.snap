// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing render, DonutChart matches snapshot 1`] = `
<div
  aria-label={null}
  aria-labelledby={null}
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-rm07xr-MuiPaper-root-MuiCard-root"
  data-mui-internal-clone-element={true}
  data-testid="savingsBox"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="MuiCardOverflow-root MuiCardOverflow-variantPlain MuiCardOverflow-colorNeutral css-17tq388-JoyCardOverflow-root"
  >
    <div
      className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        className="css-m69qwo-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          id="title"
        >
          Potensiell sparing
        </p>
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        id="numbers"
      >
         kr
      </p>
    </div>
  </div>
</div>
`;
